File: ruby/allergies/allergies.rb
# frozen_string_literal: true

class Allergies
  ALLERGENS = %w[
    eggs
    peanuts
    shellfish
    strawberries
    tomatoes
    chocolate
    pollen
    cats
  ].freeze
  private_constant :ALLERGENS
  ALLERGEN_SCORE = ->(index) { 1 << index }
  private_constant :ALLERGEN_SCORE

  attr_reader :list

  def initialize(score)
    @list = determine_allogens(score)
  end

  def allergic_to?(item)
    list.include?(item)
  end

  private

  def determine_allogens(score)
    ALLERGENS.select.with_index do |_item, index|
      (score & ALLERGEN_SCORE.call(index)).nonzero?
    end
  end
end

File: elixir/allergies/allergies.exs
defmodule Allergies do
  use Bitwise, only_operators: true

  @allergens ~w[
    eggs
    peanuts
    shellfish
    strawberries
    tomatoes
    chocolate
    pollen
    cats
  ]

  @doc """
  List the allergies for which the corresponding flag bit is true.
  """
  @spec list(non_neg_integer) :: [String.t()]
  def list(flags) do
    @allergens
    |> Enum.with_index()
    |> Enum.filter(&allergen?(&1, flags))
    |> Enum.map(&elem(&1, 0))
  end

  @doc """
  Returns whether the corresponding flag bit in 'flags' is set for the item.
  """
  @spec allergic_to?(non_neg_integer, String.t()) :: boolean
  def allergic_to?(flags, item) do
    flags
    |> list()
    |> Enum.member?(item)
  end

  defp allergen?({_item, index}, flags) do
    (flags &&& allergen_score(index)) > 0
  end

  defp allergen_score(index) do
    1 <<< index
  end
end
